src/jvm/test/org/jetbrains/kotlinx/lincheck_test/ObjectTraverserTest.kt

```diff
@@ -0,0 +1,108 @@
+ */
+class ObjectTraverserTest {
+
+    @Test
+    fun `should not traverse class and classLoader recursively while enumerating objects`() {
+        val myObject = object : Any() {
+            var clazz: Class<*>? = this::class.java
+            var classLoader: ClassLoader? = this::class.java.classLoader
+            var integer: Int = 10
+        }
+        val objectEnumeration = enumerateObjects(myObject)
+
+        Assert.assertTrue(objectEnumeration.keys.none { it is Class<*> || it is ClassLoader })
+    }
+
+    @Test
+    fun `should handle cyclic dependencies`() {
+        val objectA = object : Any() {
+            var B: Any? = null
+        }
+        val objectB = object : Any() {
+            val A: Any = objectA
+        }
+        objectA.B = objectB
+        val objectEnumeration = enumerateObjects(objectA)
+
+        Assert.assertTrue(objectEnumeration.keys.size == 2 && objectEnumeration.keys.containsAll(listOf(objectA, objectB)))
+    }
+
+    @Test
+    fun `should traverse reference-array elements for java arrays`() {
+        val o1 = Optional.of(1)
+        val o2 = Optional.of(2)
+        val o3 = Optional.of(3)
+        val myObject = object : Any() {
+            val array = arrayOf(o1, o2, o3)
+        }
+        val objectEnumeration = enumerateObjects(myObject)
+
+        Assert.assertTrue(
+            objectEnumeration.size == 8 &&
+            objectEnumeration.keys.containsAll(
+                listOf(myObject, myObject.array, *myObject.array, o1.get(), o2.get(), o3.get())
+            )
+        )
+    }
+
+    @Test
+    fun `should traverse array elements for atomic arrays`() {
+        val myObject = object : Any() {
+            val array = AtomicReferenceArray(arrayOf(1, 2, 3))
+        }
+        val objectEnumeration = enumerateObjects(myObject)
+
+        Assert.assertTrue(
+            objectEnumeration.size == 5 &&
+            objectEnumeration.keys.containsAll(
+                listOf(myObject, myObject.array, myObject.array[0], myObject.array[1], myObject.array[2])
+            )
+        )
+    }
+
+    @Test
+    fun `should traverse array elements for atomicfu arrays`() {
+        val myObject = object : Any() {
+            val array = kotlinx.atomicfu.AtomicIntArray(3)
+            init {
+                for (i in 0 until array.size) {
+                    array[i].value = i + 1
+                }
+            }
+        }
+        val objectEnumeration = enumerateObjects(myObject)
+
+        Assert.assertTrue(
+            objectEnumeration.size == 5 &&
+            objectEnumeration.keys.containsAll(
+                listOf(
+                    myObject,
+                    // atomicfu transformers [are insane and] don't allow to compile direct reference to `myObject.array`
+                    myObject.javaClass.getDeclaredField("array").get(myObject),
+                    1, 2, 3
+                )
+            )
+        )
+    }
```

Comment:
You can add a test case here for `AtomicReference<AtomicReference<...>>` case here (and similar for AFU)

Reply:
added a test with inner reference for java-atomic and atomicfu references